clients/src/main/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumer.java

```diff
@@ -75,14 +81,39 @@ public void sendRequest(Duration waitTime,
         this.callback = callback;
     }
 
-    public Set<TopicPartition> unfinishedPartitions() {
-        return unfinished;
+    private void alterSubscription(Set<TopicPartition> newSubscription) {
+        if (endOffsets.keySet().containsAll(newSubscription)) {
+            newSubscription.removeAll(endOffsets.keySet());
+        } else {
+            super.unsubscribe();
+        }
+        final HashSet<String> topics = new HashSet<>(newSubscription.size());
+        for (TopicPartition partition : newSubscription) {
+            topics.add(partition.topic());
+        }
+        super.subscribe(topics);
+    }
+
+    private void setNewOffsets(Map<TopicPartition, Long> startOffsets,
+                              Map<TopicPartition, Long> endOffsets) {
+        alterSubscription(startOffsets.keySet());
+        for (Map.Entry<TopicPartition, Long> entry : startOffsets.entrySet()) {
+            super.seek(entry.getKey(), entry.getValue());
+        }
+        this.endOffsets = endOffsets;
+        this.unfinished = startOffsets.keySet();
+    }
+
+    public void addNewOffsets(Map<TopicPartition, Long> startOffsets,
+                              Map<TopicPartition, Long> endOffsets) {
+        incomingStartOffsets.add(startOffsets);
```

Comment:
incomingStartOffsets is a List where order matters.
startOffsets and endOffsets are Map's where traversal order may be non-deterministic.

For each key in startOffsets, maybe find corresponding key from endOffsets and add its value ?